### ISSUE TITLE
Physics parameters are non-differentiable

### ISSUE THREAD

**Issue by mh-nguyen712** [deepinv/deepinv#438]
The current `update_parameters()` of `Physics` and `NoiseModel`  does not allow one to differentiate the output of the physics w.r.t to the parameter. For example, the following code snippet won't work:
```
physics = dinv.physics.Blur(device='cpu')
filter = torch.rand((1, 1, 5, 5))
filter /= filter.sum(dim=(-2,-1), keepdim=True)

with torch.enable_grad():
    filter.requires_grad_(True)
    y_cur = physics.A(x, filter=filter)
    loss = torch.nn.functional.mse_loss(y_cur, y)
    
grad = torch.autograd.grad(loss, filter)[0]
```

The reason is inside the `update_parameters()`, we forced the `filter` to be a `nn.Parameter` with `requires_grad = False`.

A simple work-around could be:
```
  def update_parameters(self, **kwargs):
      if kwargs:
          for key, value in kwargs.items():
              if (
                  value is not None
                  and hasattr(self, key)
                  and isinstance(value, torch.Tensor)
              ):
                  self.register_buffer(key, value)
```
The `register_buffer` acts similar to `nn.Parameter`, except that `filter` won't be show up in `physics.parameters()`, but I think it's not very important. The `filter` will be in the `state_dict` . Methods like `physics.to(...)` or `physics.cuda()`  will work properly.

This is related to #412  #387 #414 #383

I can create a PR soon for adressing this issue.

**Comment by Andrewwango** [comment]
Can we add them as sub-issues?

**Comment by tachella** [comment]
Many thanks Hai, fixing this would be super interesting!

I have a doubt regarding your proposed solution: it would be nice to optimize over parameters with a simple optimizer such as `Adam`, I think that the proposed solution won't work in this case because `filter` won't be in the `physics.parameters()`?

**Comment by mh-nguyen712** [comment]
We can optimize the param as follows:
```
filters = torch.tensor(..., requires_grad=True)
optimizer = Adam([filters], ...)
for _ in ...:
   optimizer.zero_grad()
    loss = loss_fn(physics(..., filters), ...)
    loss.backward()
    optimizer.step()
``
The physics only keeps track the gradient w.r.t to param if the users set `requires_grad = True` for the initial param. Otherwise, it does not.

**Comment by tachella** [comment]
A naive question: why not adding the option to the `Physics` class (e.g. `learn_params`) for setting `requires_grad=True` ?

**Comment by mh-nguyen712** [comment]
> A naive question: why not adding the option to the `Physics` class (e.g. `learn_params`) for setting `requires_grad=True` ?

This way we leave the users choose whenever they want to differentiate the param (or something else) but the physics itself is always differentiable (as in torch). And `register_buffer` makes sure that all physics-dependent params will work properly with methods like `.to()` or `.cuda()`, ... This way we dont need the extra variable `learn_params`.

**Comment by tachella** [comment]
Is there a way to still save parameters when we do `torch.save(physics.state_dict())`?

**Comment by mh-nguyen712** [comment]
> Is there a way to still save parameters when we do `torch.save(physics.state_dict())`?

Yes! When we save a param with `self.register_buffer()`, it will appear in the state dict.

**Comment by tachella** [comment]
ok, then it looks good to me!
